### ISSUE TITLE
Ruby 2.5.0, fastly-1.15.0:  Purge Errors

### ISSUE THREAD

**Issue by tyrauber** [fastly/fastly-ruby#130]
In upgrading our Rails App, I also updated the fastly-rails gem from 0.2 to 0.8.0. But in doing so, I am  now seeing a whole string of new failures related to the fastly-rails gem.  Roughly ~800 failures in a 24 hour period from a single worker. We also updated Ruby from 2.3.1 to 2.5.0.  Perhaps, related?

Environment:

- Rails 5.1.5
- Ruby 2.5.0
- fastly-rails 0.8.0
- fastly-ruby 1.15.0

Here is the assortment of errors we are seeing with relevant stacktraces.

Anyone else seeing anything like this?  Any suggestions?

I posted this issue to [fastly-rails](https://github.com/fastly/fastly-rails/issues/80), but looking at the stacktrace, it looks like it may actually be a fastly-ruby issue.

- **JSON::ParserError: Empty input at line 1, column 1 [parse.c:926] in 'HTTP/1.1 200 OK Content-Type: application/json C**
```
1 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 124 in parse
2 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 124 in post_and_put
3 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 87 in post
4 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/service.rb" line 74 in purge_by_key
5 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/client.rb" line 12 in purge_by_key
6 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails.rb" line 26 in purge_by_key
7 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/active_record/surrogate_key.rb" line 37 in purge
```

- **Net::HTTPBadResponse: wrong status line: "ntent-Length: 50”**
```
1 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 42 in read_status_line
2 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 29 in read_new
3 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1494 in block in transport_request
4 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1491 in catch
5 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1491 in transport_request
6 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1464 in request
16 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1478 in send_entity
17 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1266 in post
18 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 122 in post_and_put
19 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 87 in post
20 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/service.rb" line 74 in purge_by_key
21 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/client.rb" line 12 in purge_by_key
22 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails.rb" line 26 in purge_by_key
23 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/active_record/surrogate_key.rb" line 37 in purge
```
- **Errno::EBADF: Bad file descriptor**
```
1 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 181 in wait_readable
2 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 181 in rbuf_fill
3 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 125 in read
4 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 293 in block in read_body_0
5 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 278 in inflater
6 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 283 in read_body_0
7 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 204 in read_body
8 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1479 in block in send_entity
9 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1503 in block in transport_request
10 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 165 in reading_body
11 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1502 in transport_request
12 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1464 in request
```
- I**OError: closed stream**
```
1 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 181 in wait_readable
2 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 181 in rbuf_fill
3 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 157 in readuntil
4 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/protocol.rb" line 167 in readline
5 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 40 in read_status_line
6 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http/response.rb" line 29 in read_new
7 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1494 in block in transport_request
8 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1491 in catch
9 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1491 in transport_request
10 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1464 in request
20 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1478 in send_entity
21 File "/app/vendor/ruby-2.5.0/lib/ruby/2.5.0/net/http.rb" line 1266 in post
22 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 122 in post_and_put
23 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/client.rb" line 87 in post
24 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-1.15.0/lib/fastly/service.rb" line 74 in purge_by_key
25 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/client.rb" line 12 in purge_by_key
26 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails.rb" line 26 in purge_by_key
27 File "/app/vendor/bundle/ruby/2.5.0/gems/fastly-rails-0.8.0/lib/fastly-rails/active_record/surrogate_key.rb" line 37 in purge
```

**Comment by tyrauber** [comment]
In digging into the code and stacktraces, it seems like these errors are caused by intermittent invalid responses.  The code appears to be expecting one format and receiving another and is unable to parse the result.

All the errors seem to originate from these three lines:
https://github.com/fastly/fastly-ruby/blob/master/lib/fastly/client.rb:122-124
```
resp  = http.send(method, path, query, headers(extras).merge('Content-Type' =>  'application/x-www-form-urlencoded'))
fail Error, resp.body unless resp.kind_of?(Net::HTTPSuccess)
JSON.parse(resp.body)
```

The JSON::ParseError is related to 124, which means that resp.body is not valid json.

The Net::HTTPBadResponse Error is related to line 123. The resp is not resp.kind_of?(Net::HTTPSuccess).

IOError: closed stream Error is related to line 122. Perhaps, http.send is timing out, or headers.merge isn't resulting in the right Content-Type?

Errno::EBADF: Bad file descriptor is also related to line 122.

What's crazy is these errors are intermittent. I am not sure why one request would result in a valid response, while another would be invalid. Definitely strange.

**Comment by palin** [comment]
Using fastly gem v.2.0.0 I've got similar issue.
I'm purging a url which points to a png image.
I call `fastly.purge(url, true)`, then I get this:

```
JSON::ParserError: 757: unexpected token at '<!DOCTYPE html>
<html>
  <head>
    <title>200 Ban added</title>
  </head>
  <body>
    <h1>Error 200 Ban added</h1>
    <p>Ban added</p>
    <h3>Guru Meditation:</h3>
    <p>XID: 582400232</p>
    <hr>
    <p>Varnish cache server</p>
  </body>
</html>
'
```

So fastly code tries to parse JSON response, but the response is HTML, so it blows up.
Here's the line which blows up:
https://github.com/fastly/fastly-ruby/blob/master/lib/fastly/client.rb#L114

I can only manually rescue the JSON::ParserError.

**Comment by tyrauber** [comment]
@palin Check out the PR #131 and see if it resolves the issue. I don't believe the JSON response is being used, so I think it is safe to rescue the JSON.parse and return the raw resp.body

**Comment by tyrauber** [comment]
I have debugged the issue and confirmed that all errors reported above are a result of the gems utilization / implementation of Ruby's net::http.  I reverted back to fastly 1.1.4 and all errors are resolved.

Looking through the version history, I noticed that  [Version 1.2.0](https://github.com/fastly/fastly-ruby/commit/67803cff40e4086e85d132cca028c5c09e0b0d46) replaced the Curl implementation with Net::Http.  While this change may have resulted in better error handling it produced a significant number of errors in our production environment.

This leaves the open question of whether the curl implementation in 1.1.4 and below is silently erring?  Are purge requests failing in 1.1.4 and earlier and just not reporting the error? 

I guess I could fork 1.15.0 and rescue the errors, but that doesn't answer the question as to why net::http is erring out.  I suspect it has something to do with invalid responses being intermittently returned by Fastly on cache purge.

**Comment by ziggythehamster** [comment]
This is also happening with us, and we occasionally get EBADF as well. It seems like someone is reading a stream, but isn't reading it to completion, and isn't checking to see if the stream is still valid when reading. There are lots of edge cases with Net::HTTP like this, and is part of the reason we use Typhoeus instead.

**Comment by thommahoney** [comment]
@tyrauber @ziggythehamster Are either of you using this gem in a multi-threaded environment like Sidekiq or Puma? If so, you might see issues like this. The suggestion would be to use a ConnectionPool so that only one thread is utilizing the Fastly client at a time. See https://github.com/mperham/sidekiq/wiki/Advanced-Options#connection-pooling for details.

If you aren't using the gem in a multi-threaded environment, let us know.

**Comment by ziggythehamster** [comment]
Yeah, we absolutely are. Are you saying this gem isn't threadsafe? We can wrap our calls to the Fastly gem in a Mutex (a pool doesn't really work because we're using fastly-rails and not directly calling Fastly)

**Comment by tyrauber** [comment]
That's a bingo. All three here. Fastly-Rails, Puma and Sidekiq. @thommahoney,  I assume even If we didn't push the purge requests to Sidekiq, we'd still see the errors due to Puma? 

What's preventing the gem from being threadsafe?

**Comment by thommahoney** [comment]
> What's preventing the gem from being threadsafe?

I guess it'd be this line mostly: https://github.com/fastly/fastly-ruby/blob/71b0dee9e78119fb70714a4d5343e0cc545932e7/lib/fastly/client.rb#L24

The `@http` instance variable is used within the `post_and_put` method (via the `http` accessor), and if you're invoking that codepath from different threads within the same process, then the net/http instance can be in an unpredictable state (eg. finishing the previous request)

One solution to this would be to make `purge_by_key` use a new instance of `Net::HTTP` each time it is invoked. This has performance tradeoffs (memory overhead, connection re-establishment (?), etc) so I'd prefer not to add it to this gem.

Perhaps a mutex or connection pool could be added to the fastly-rails implementation here: https://github.com/fastly/fastly-rails/blob/6ed13a694b8753d43515079ae667fe6f8c0fcd23/lib/fastly-rails.rb#L29-L37

**Comment by ziggythehamster** [comment]
Well, you only need one client per thread, right? So you could do the "build a client every time" method but use a ThreadLocalVar to hold on to it. Then, subsequent requests in the same thread for a client don't build a new client and you're threadsafe. Of course, if you've got a ton of threads, this doesn't scale as well as a connection pool does (unless you make the argument that every thread needs to concurrently access a client, though I doubt that's a good architecture since we're talking about purging here), but a connection pool is also a larger refactoring and might involve API changes, and I don't really care about allocating an extra 10 objects which each occupy a few KB once. People on a more restrictive platform like Heroku probably have a different opinion.

TL;DR: I'd create the client as a ThreadLocalVar as well as any other state tied to a client instance and call this fixed

**Comment by thommahoney** [comment]
@ziggythehamster @tyrauber I'm not sure if this works yet but I opened a PR here: https://github.com/fastly/fastly-ruby/pull/139

**Comment by johvet** [comment]
Any chance that PR could be reviewed (or ideally revised in regards the use of `Concurrent::ThreadLocalVar`)

**Comment by demersus** [comment]
The alternative approach with `Concurrent::ThreadLocalVar` is here: #144

**Comment by thommahoney** [comment]
I have released v2.3.0 which implements separate instances of `net/http` per thread. If `concurrent-ruby` is available in the runtime, it will use a `Concurrent::ThreadLocalVar` instead. Please comment here or open a new issue if this new release causes any trouble. My tests (which included purging 10 URLs in parallel) were successful with and without concurrent-ruby loaded.
